pkg/clients/bucketpolicy/bucketpolicy.go

```diff
@@ -0,0 +1,136 @@
+func GenerateBucketPolicyInstance(in v1alpha1.BucketPolicyParameters, ck *storage.Policy) {
+	ck.Bindings = make([]*storage.PolicyBindings, len(in.Policy.Bindings))
+	for i, v := range in.Policy.Bindings {
+		ck.Bindings[i] = &storage.PolicyBindings{}
+		if v.Condition != nil {
+			ck.Bindings[i].Condition = &storage.Expr{
+				Description: v.Condition.Description,
+				Expression:  v.Condition.Expression,
+				Location:    v.Condition.Location,
+				Title:       v.Condition.Title,
+			}
+		}
+		ck.Bindings[i].Members = make([]string, len(v.Members))
+		copy(ck.Bindings[i].Members, v.Members)
+		ck.Bindings[i].Role = v.Role
+	}
+	ck.Version = policyVersion
+}
+
+// IsUpToDate checks whether current state is up-to-date compared to the given
+// set of parameters.
+func IsUpToDate(in *v1alpha1.BucketPolicyParameters, observed *storage.Policy) (bool, error) {
+	generated, err := copystructure.Copy(observed)
+	if err != nil {
+		return true, errors.Wrap(err, errCheckUpToDate)
+	}
+	desired, ok := generated.(*storage.Policy)
+	if !ok {
+		return true, errors.New(errCheckUpToDate)
+	}
+	GenerateBucketPolicyInstance(*in, desired)
+	return cmp.Equal(desired, observed, cmpopts.EquateEmpty(),
+		cmpopts.IgnoreFields(storage.Policy{}, "Version"),
+		cmpopts.SortSlices(func(i, j *storage.PolicyBindings) bool { return i.Role > j.Role }),
+		cmpopts.SortSlices(func(i, j string) bool { return i > j })), nil
+}
+
+// IsEmpty returns if Policy is empty
+func IsEmpty(in *storage.Policy) bool {
+	return in.Bindings == nil
+}
+
+// BindRoleToMember updates *storage.Policy instance with BucketPolicyMemberParameters.
+// returns true if policy changed
+func BindRoleToMember(in v1alpha1.BucketPolicyMemberParameters, ck *storage.Policy) bool {
+	ck.Version = policyVersion
+	for _, b := range ck.Bindings {
+		if b.Role == in.Role {
+			for _, m := range b.Members {
+				if m == gcp.StringValue(in.Member) {
+					// role already bound to member, no change
+					return false
+				}
+			}
+			// role already exist, add member
+			b.Members = append(b.Members, gcp.StringValue(in.Member))
+			return true
+		}
+	}
+	// role does not exist, add binding with role and member
+	ck.Bindings = append(ck.Bindings, &storage.PolicyBindings{
+		Role:    in.Role,
+		Members: []string{gcp.StringValue(in.Member)},
+	})
+	return true
+}
+
+// UnbindRoleFromMember generates *storage.Policy instance from BucketPolicyMemberParameters.
+// returns true if bound (i.e. policy changed)
+func UnbindRoleFromMember(in v1alpha1.BucketPolicyMemberParameters, ck *storage.Policy) bool {
+	for _, b := range ck.Bindings {
+		if b.Role == in.Role {
+			ix := -1
+			for i, m := range b.Members {
+				if m == gcp.StringValue(in.Member) {
+					// found member binding in role
+					ix = i
+					break
+				}
+			}
+			if ix >= 0 {
+				// remove member located at index ix
+				b.Members = append(b.Members[:ix], b.Members[ix+1:]...)
+				return true
```

Comment:
I noticed we return early here, I am guessing there would not be the case where a subsequent binding would also need the member removed?

Reply:
yes, policy member CR sets one role to one member (no list in spec), so, once we found it we are done.